sdks/python/apache_beam/pipeline.py

```diff
@@ -396,28 +396,40 @@ def replace_all(self, replacements):
     for override in replacements:
       self._check_replacement(override)
 
-  def run(self, test_runner_api=True):
-    """Runs the pipeline. Returns whatever our runner returns after running."""
+  def run(self, test_runner_api=True, runner=None, options=None):
+    """Runs the pipeline. Returns whatever our runner returns after running.
+
+    If another runner instance and options are provided, that runner will
+    execute the pipeline with the given options. If either of them is not set,
+    the default runner will run the pipeline with the original options
+    assigned to the pipeline. The usage is similar to directly invoking
+    `runner.run_pipeline(pipeline, options)`.
+    """
+    runner_in_use = self.runner
+    options_in_use = self._options
+    if runner and options:
```

Comment:
What if either runner or options are not provided? Should that throw an error? Currently, if only one is provided, it'll be ignored - and that would be quite surprising for users.

Reply:
You're right! This will surprise the user. I've changed it to throw error if either is not provided instead of ignoring the input by default.